### PR TITLE
Balance Enforcement Invariant

### PR DESCRIPTION
**Pull request by epociask**
## Fixes Issue

Fixes #
## Changes proposed

Added:
* Address balance oracle that tracks native ETH amounts for some address
* Addressing optionality for invariants and pipelines
* Balance enforcement invariant using addressing functionality
* Alerting subsystem with event loop
* Ability to post alerts to slack for balance enforcement alerts

### CODE REVIEW

File: internal/api/models/invariant.go [hunk 1]
```diff
@@ -7,56 +7,88 @@ import (
...
+type InvResult = map[string]string
+
+// InvResponse ... Response for invariant operation request
+type InvResponse struct {
+	Code   int               `json:"status_code"`
+	Status InvResponseStatus `json:"status"`
+
+	Result InvResult `json:"result"`
+	Error  string    `json:"error"`
+}
```
Comment (a-drain) [inline internal/api/models/invariant.go:94]: Are we not planning to use gRPC for creating and maintaining rpc endpoints? Defining endpoints via protobufs could help with strictly typing endpoint request/response models as well as autogenerating code like this

File: internal/config/config.go [hunk 2]
```diff
@@ -22,13 +22,18 @@ const (
 	Development Env = "development"
 	Production  Env = "production"
 	Local       Env = "local"
+
+	trueEnvVal = "1"
```
Comment (a-drain) [inline internal/config/config.go:26]: what is this?

File: internal/core/constants.go [hunk 3]
```diff
@@ -55,13 +74,67 @@ func (it InvariantType) String() string {
...
+func (ad AlertDestination) String() string {
+	switch ad {
+	case Slack:
+		return "slack"
+	case CounterParty:
+		return "counterparty"
+	default:
+		return "unknown"
+	}
+}
```
Comment (a-drain) [inline internal/core/constants.go:119]: Bit confused as to what a counterparty is. Assuming it just means anything but slack?

File: internal/engine/registry/invocation.go [hunk 4]
```diff
@@ -2,6 +2,7 @@ package registry
```
Comment (a-drain) [inline internal/engine/registry/invocation.go:2]: Is there documentation for the state of each invariant? Would probably be good to at least add high level docs on what each invariant does adjacent to the implementation of the invariant or within the docs somewhere.

File: internal/engine/store.go [hunk 5]
```diff
@@ -81,3 +83,12 @@ func (ss *sessionStore) RemoveInvSession(_ core.InvSessionUUID,
...
 	return nil
 }
+
+func TestSessionStore(t *testing.T) {
+	// Setup
+	ss := NewSessionStore()
+
+	// Test GetInvSessionByUUID
+	_, err := ss.GetInvSessionByUUID(core.NilInvariantUUID())
+	assert.Error(t, err, "should error")
```
Comment (a-drain) [inline internal/engine/store.go:93]: why not create store_test.go

File: internal/state/memory.go [hunk 6]
```diff
@@ -0,0 +1,65 @@
+package state
```
Comment (a-drain) [inline internal/state/memory.go:1]: In mem store is a good interim solution but we should definitely have a plan to move away from this and into some sort of persistent storage solution. The in mem store is going to require us to keep vertically scaling the instance used to run this, not to mention potentially exceeding the memory limit causing the system to crash

**Review by a-drain — COMMENTED** [review]
Left a few minor questions, otherwise looks good!

File: internal/state/memory.go (hunk 6 above)
Comment (epociask) [reply internal/state/memory.go:1]: Agreed, this is called out in documentation as well

File: internal/api/models/invariant.go (hunk 1 above)
Comment (epociask) [reply internal/api/models/invariant.go]: Agreed but think that's ultimately out of scope for now

**Review by anupsv-cb — APPROVED** [review]

**Pull request merged**